attic/sisku.1
.\" Automatically generated by Pod::Man 2.12 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SISKU 1"
.TH SISKU 1 "2009-06-10" "" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
\&\fBsisku\fR \- search for \fIvalsi\fR in the official Lojban \fIgismu\fR & \fIcmavo\fR lists
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBsisku\fR [\fB\-aCcFGgiqRrw\fR] \fIvalsi\fR ...
.PP
\&\fBsisku\fR [\fB\-dek\fR] [\fB\-aCcFGgiqw\fR] \fIdefinition\fR ...
.PP
\&\fBsisku\fR [\fB\-CcGg\fR]
.PP
\&\fBsisku\fR \fB\-p\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBsisku\fR searches through the official \fIgismu\fR list and/or the official
\&\fIcmavo\fR list and prints out the first entry that matches each operand.  It can
search based on \fIvalsi\fR, \fIrafsi\fR, keyword, or definition using either plain
text comparison or \fIperl\fR\|(1) regular expressions.  A separate search is
performed for each operand, and the results for each are printed out separately
(i.e., not sorted or merged together).  The exact way in which arguments are
matched against text is described below under \*(L"\s-1PATTERN\s0 \s-1MATCHING\s0\*(R".
.PP
When no strings to search for are given, \fBsisku\fR prints out a random \fIvalsi\fR
(by default, an entry from the \fIgismu\fR list, though this can be changed by
specifying one of the \fB\-CcG\fR switches) and then exits.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-a\fR" 4
.IX Item "-a"
Print out all matching entries for each operand rather than just the first
match for each.  Note that each operand will still be searched for separately,
and the results will not be sorted or merged together.
.IP "\fB\-C\fR" 4
.IX Item "-C"
Retrieve only entries from the \fIcmavo\fR list, \fBexcluding\fR compound \fIcmavo\fR.
\&\fIrafsi\fR will not be output unless the \fB\-r\fR or \fB\-R\fR option is also given.
.IP "\fB\-c\fR" 4
.IX Item "-c"
Retrieve only entries from the \fIcmavo\fR list, \fBincluding\fR compound \fIcmavo\fR.
\&\fIrafsi\fR will not be output unless the \fB\-r\fR or \fB\-R\fR option is also given.
.IP "\fB\-d\fR" 4
.IX Item "-d"
Search English definitions for matches.
.IP "\fB\-e\fR" 4
.IX Item "-e"
Search through both English keywords and definitions for matches.
.IP "\fB\-F\fR" 4
.IX Item "-F"
Allow non-regex operands to match strings that contain them rather than just
strings that equal them, i.e., make them \*(L"free floating.\*(R"  Note that this
option has no effect on matching of definitions, as those matches are always
\&\*(L"free floating\*(R" unless the user uses \f(CW\*(C`^\*(C'\fR or \f(CW\*(C`$\*(C'\fR.
.IP "\fB\-G\fR" 4
.IX Item "-G"
Retrieve only \fIgismu\fR.
.IP "\fB\-g\fR" 4
.IX Item "-g"
Retrieve only entries from the \fIgismu\fR list, i.e., both \fIgismu\fR and any
\&\fIcmavo\fR that have \fIrafsi\fR.
.IP "\fB\-i\fR" 4
.IX Item "-i"
Search case-insensitively.
.IP "\fB\-k\fR" 4
.IX Item "-k"
Search English keywords for matches.
.IP "\fB\-p\fR" 4
.IX Item "-p"
Print out the paths of the \fIgismu\fR and \fIcmavo\fR lists that \fBsisku\fR searches.
If the \fB\-c\fR or \fB\-C\fR option is specified, only the path of the \fIcmavo\fR list
will be printed, and likewise for \fB\-g\fR or \fB\-G\fR with the \fIgismu\fR list.
.IP "\fB\-q\fR" 4
.IX Item "-q"
Print out the quantity of matches for each operand rather than the matches
themselves.
.IP "\fB\-R\fR" 4
.IX Item "-R"
Attempt to match \fBonly\fR \fIrafsi\fR and display only those \fIvalsi\fR that have at
least one matching \fIrafsi\fR.
.IP "\fB\-r\fR" 4
.IX Item "-r"
Also attempt to match \fIrafsi\fR and display any \fIvalsi\fR that has at least one
matching \fIrafsi\fR.
.IP "\fB\-w\fR" 4
.IX Item "-w"
Only display matches that match on a full word, i.e., act as if there is a
\&\f(CW\*(C`\eb\*(C'\fR at either end of each operand.
.Sh "Handling Mutually Exclusive Options"
.IX Subsection "Handling Mutually Exclusive Options"
Due to Getopt::Std's parsing of all options at once, in contrast to the \*(L"one at
a time\*(R" method of \fIgetopt\fR\|(3), mutually exclusive options cannot be resolved
by checking which one came last; instead, an arbitrary set of precedences is
used:
.IP "\(bu" 4
The \fB\-C\fR and \fB\-c\fR options override the \fB\-G\fR and \fB\-g\fR options.
.IP "\(bu" 4
\&\fB\-C\fR, \fB\-G\fR, and \fB\-R\fR override \fB\-c\fR, \fB\-g\fR, and \fB\-r\fR, respectively.
.IP "\(bu" 4
If both \fB\-d\fR and \fB\-k\fR are given, they are treated as if just \fB\-e\fR had been
supplied.
.IP "\(bu" 4
If \fB\-e\fR and either \fB\-d\fR or \fB\-k\fR are given, the \fB\-e\fR is ignored.
.IP "\(bu" 4
If \fB\-d\fR, \fB\-e\fR, or \fB\-k\fR is given, \fB\-r\fR and \fB\-R\fR are ignored.
.IP "\(bu" 4
The \fB\-q\fR option overrides the \fB\-a\fR option.
.SH "PATTERN MATCHING"
.IX Header "PATTERN MATCHING"
Treatment of operands is handled using a rather straightforward rule: if an
operand contains any special regex characters, it is treated as a Perl regular
expression; otherwise, it is treated as a literal string that the text being
compared against must be equal to or (if the \fB\-F\fR option is given or
definitions are being searched) must contain.  As an exception, parentheses are
not recognized as special characters because, due to the lack of capturing in
\&\fBsisku\fR, they serve no use unless other special characters are present.  For
reference, the special regex characters recognized are:
.PP
.Vb 1
\&    . * + ? ^ $ | [ ] { } \e
.Ve
.PP
If you wish to include one of these characters in a search pattern without
giving it any special meaning, you must precede it with a backslash (possibly
two backslashes, depending on your shell & use of quotes).  For more
information on \fIperl\fR\|(1)'s regular expressions, run \f(CW\*(C`perldoc perlrequick\*(C'\fR,
\&\f(CW\*(C`perldoc perlretut\*(C'\fR, and/or \f(CW\*(C`perldoc perlre\*(C'\fR.  If you supply a bad regex to
\&\fBsisku\fR, whatever happens is your own fault.
.PP
When not using any of the \fB\-dek\fR options, 'h's in the operands are converted
to apostrophes before searching.
.PP
Note that any periods at the beginning of \fIcmavo\fR in the \fIcmavo\fR list are
ignored both when using regexes & when using literal strings.  Thus, the
operand \f(CW\*(C`.e\*(C'\fR will match \fIbe\fR but not \fI.e\fR, and forcing literal
interpretation with \f(CW\*(C`\e.e\*(C'\fR will cause the query to return no results.  Periods
elsewhere in \fIcmavo\fR are left intact, but note that \fIna.a\fR is the only entry
in the official \fIcmavo\fR list to have a non-leading period.
.SH "RESTRICTIONS"
.IX Header "RESTRICTIONS"
\&\fBsisku\fR does not search the notes or hints sections of entries.
.PP
Some \fIcmavo\fR which appear in both the \fIgismu\fR list and the \fIcmavo\fR list have
different definitions in each, and currently the best way to handle them is to
concatenate them with "\f(CW\*(C` \-\-\- \*(C'\fR".  This can generate odd-looking results,
especially for \fIcmavo\fR whose two definitions differ only in a single word.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIjvozba\fR\|(1), Lojban::Valsi, Lojban::Vlasisku
.PP
\&\fIThe Complete Lojban Language\fR by John Woldemar Cowan
.PP
<http://www.lojban.org/>
.SH "AUTHOR"
.IX Header "AUTHOR"
John T. \*(L"kamymecraijun.\*(R" Wodder \s-1II\s0 <minimiscience@gmail.com>
.SH "LICENSE"
.IX Header "LICENSE"
Feel free to do whatever the \fIbais.\fR you want with this.


perl/CHANGES.txt
CHANGES TO THE PERL CODE SINCE JBOVAL 1.1
- Lojban::Valsi now supports {cmene}, {lujvo}, & {fu'ivla}.  The names of these
  types may be returned by Lojban::Valsi::klesi(), and there are xuXXX()
  methods for each of them. *
- Renamed Lojban::Vlatai::terjvo() to "veljvo()"
- Lojban::Vlatai: Renamed $c, $v, and $lidne to $C, $V, and $CC, respectively. *
- Made the Lojban::Vlatai regexes & functions case-insensitive. *
- Added the following functions:
 - Lojban::Vlatai:
  - xugismu *
  - xucmavo *
  - xucmene *
  - xulujvo *
  - xufuhivla *
  - jvozba (along with the constants JVOZBA_BRIVLA and JVOZBA_CMENE) *
  - xulujmaho *
  - fadgau *
 - Lojban::Valsi:
  - cnino *
  - veljvo *
  - fukpi *
  - cmuselmaho *
  - catniho *
  - klesi *
  - krarafsi *
  - termre *
  - xuvladra *
  - fadni *
  - xubrivla *
  - xusuzmaho *
- Added the $Lojban::Vlatai::dotside configuration variable (not exported) *
- Added the :lujvo and :klesi export tags to Lojban::Vlatai *
- Lojban::Vlatai::jvomre can now be passed the {lujvo}'s {rafsi} after the
  {lujvo} itself so that jvomre doesn't have to split it. *
- Added the following regexes to Lojban::Vlatai:
 - $Vy *
 - $CyC *
 - $C_C *
 - $CxC *
- Gave sisku.pl & jvozba.pl --help messages
- Added @Lojban::Vlatai::klesi, a list of all recognized {vlalei} *
- Gave compound {cmavo} their own {vlalei}, "lujma'o"
- @Lojban::Valsi::{qw< voiced unvoiced nodup verboten >} are now gone.
- &Lojban::Vlatai::vlalei now returns undef for unrecognized word forms *
- Lojban::Vlatai's :fancu export tag is now gone, as it was the same as
  :DEFAULT
- Got rid of $Lojban::Vlasisku::dosyste
- Lojban::Valsi: The isXXX() methods are now named isxxx() *
- Lojban::Valsi::printEntry() is now named "cusku()" and treats calculated
  attributes ... specially *
- getValsi() matches against a word's normalized form, not against the {valsi}
  itself. *
- Added Lojban::Vlatid *
- The Lojban::Vlasisku functions now take an optional first argument giving the
  Lojban::Vlatid object to read {valsi} from; if it is omitted, a global Vlatid
  object is used instead. *
- Lojban::Vlasisku has been completely rewritten. *

* These still need to be documented in the POD.


jboval-1.1/sisku.pl
#!/usr/bin/perl -w
# $Id$
use strict;
use Getopt::Std;
use Lojban::Valsi;
use Lojban::Vlasisku qw< :DEFAULT :stodi >;

my %opts;
getopts('cgGeiwFarRCqkdp', \%opts) || exit 2;

# Eliminate conflicting options:
$opts{g} = $opts{G} = 0 if $opts{c} || $opts{C};
$opts{g} = 0 if $opts{G};
$opts{c} = 0 if $opts{C};
$opts{r} = $opts{R} = 0 if $opts{e};
$opts{r} = 0 if $opts{R};
$opts{e} = 1 if $opts{k} || $opts{d};
$opts{k} = $opts{d} = 0 if $opts{k} && $opts{d};

if ($opts{p}) {
 print $Lojban::Vlasisku::gimste, "\n" unless $opts{c} || $opts{C};
 print $Lojban::Vlasisku::mahoste, "\n" unless $opts{g} || $opts{G};
 exit 0;
}

my($wordSearch, $keySearch, $defSearch) = (\&getValsi, \&getValsiByKeyword,
 \&getValsiByDefinition);

if ($opts{c} || $opts{C}) {
 $wordSearch = \&getCmavo;
 $keySearch = \&getCmavoByKeyword;
 $defSearch = \&getCmavoByDefinition;
} elsif ($opts{g} || $opts{G}) {
 $wordSearch = \&getGismu;
 $keySearch = \&getGismuByKeyword;
 $defSearch = \&getGismuByDefinition;
}

my $patopts;
$patopts = VLASISKU_INSENSITIVE if $opts{i};
if ($opts{w}) {
 if (defined $patopts) { $patopts |= VLASISKU_WHOLE_WORD }
 else { $patopts = VLASISKU_WHOLE_WORD }
}
if ($opts{F} && !defined $patopts) { $patopts = 0 }

if (@ARGV) {
 foreach (@ARGV) {
  my @valsi;
  my $op = defined $patopts && !/[][.*+?^\$\\|{}]/o ? $patopts | ($opts{F} ? 0
   : VLASISKU_ANCHORED) | VLASISKU_LITERAL : $patopts;
  if ($opts{e}) {
   @valsi = &$keySearch($_, $op) unless $opts{d};
   push @valsi, &$defSearch($_, defined $op ? ($op & ~VLASISKU_ANCHORED) : $op)
    unless $opts{k};
   @valsi = Lojban::Valsi::girxre @valsi unless $opts{k} || $opts{d};
  } else {
   tr/h/'/;
   @valsi = getValsiByRafsi($_, $op) if $opts{r} || $opts{R};
   push @valsi, &$wordSearch($_, $op) unless $opts{R};
   @valsi = Lojban::Valsi::girxre @valsi if $opts{r};
   @valsi = grep { $_->isCmavo() } @valsi if ($opts{c} || $opts{C}) &&
    ($opts{r} || $opts{R});
  }
  @valsi = grep { $_->isGismu() } @valsi if $opts{G};
  @valsi = grep { !$_->isCompoundCmavo() } @valsi if $opts{C};
  @valsi = ($valsi[0]) if !($opts{a} || $opts{q}) && @valsi;
  # Yes, getting the whole set and then discarding all but one is bad, but it's
  # the easiest way to implement the -G, -C, and -c with -[rR] options without
  # rewriting Vlasisku.pm.
  if ($opts{q}) { print scalar @valsi, "\n" }
  else { printEntry $_ for @valsi }
 }
} else {
 if ($opts{c}) {
  my $cmavo = cunmaho();
  $cmavo->printEntry();
 } elsif ($opts{C}) {
  my $cmavo = cunmaho();
  $cmavo = cunmaho() while $cmavo->isCompoundCmavo();
  $cmavo->printEntry();
 } elsif ($opts{G}) {
  my $gismu = cungihu();
  $gismu = cungihu() until $gismu->isGismu();
  $gismu->printEntry();
 } else {
  my $gimstecmi = cungihu();
  $gimstecmi->printEntry();
 }
}

__END__

=pod

=head1 NAME

B<sisku> - search for I<valsi> in the official Lojban I<gismu> & I<cmavo> lists

=head1 SYNOPSIS

B<sisku> [B<-aCcFGgiqRrw>] I<valsi> ...

B<sisku> [B<-dek>] [B<-aCcFGgiqw>] I<definition> ...

B<sisku> [B<-CcGg>]

B<sisku> B<-p>

=head1 DESCRIPTION

B<sisku> searches through the official I<gismu> list and/or the official
I<cmavo> list and prints out the first entry that matches each operand.  It can
search based on I<valsi>, I<rafsi>, keyword, or definition using either plain
text comparison or L<perl(1)> regular expressions.  A separate search is
performed for each operand, and the results for each are printed out separately
(i.e., not sorted or merged together).  The exact way in which arguments are
matched against text is described below under L</"PATTERN MATCHING">.

When no strings to search for are given, B<sisku> prints out a random I<valsi>
(by default, an entry from the I<gismu> list, though this can be changed by
specifying one of the B<-CcG> switches) and then exits.

=head1 OPTIONS

=over

=item B<-a>

Print out all matching entries for each operand rather than just the first
match for each.  Note that each operand will still be searched for separately,
and the results will not be sorted or merged together.

=item B<-C>

Retrieve only entries from the I<cmavo> list, B<excluding> compound I<cmavo>.
I<rafsi> will not be output unless the B<-r> or B<-R> option is also given.

=item B<-c>

Retrieve only entries from the I<cmavo> list, B<including> compound I<cmavo>.
I<rafsi> will not be output unless the B<-r> or B<-R> option is also given.

=item B<-d>

Search English definitions for matches.

=item B<-e>

Search through both English keywords and definitions for matches.

=item B<-F>

Allow non-regex operands to match strings that contain them rather than just
strings that equal them, i.e., make them "free floating."  Note that this
option has no effect on matching of definitions, as those matches are always
"free floating" unless the user uses C<^> or C<$>.

=item B<-G>

Retrieve only I<gismu>.

=item B<-g>

Retrieve only entries from the I<gismu> list, i.e., both I<gismu> and any
I<cmavo> that have I<rafsi>.

=item B<-i>

Search case-insensitively.

=item B<-k>

Search English keywords for matches.

=item B<-p>

Print out the paths of the I<gismu> and I<cmavo> lists that B<sisku> searches.
If the B<-c> or B<-C> option is specified, only the path of the I<cmavo> list
will be printed, and likewise for B<-g> or B<-G> with the I<gismu> list.

=item B<-q>

Print out the quantity of matches for each operand rather than the matches
themselves.

=item B<-r>

Also attempt to match I<rafsi> and display any I<valsi> that has at least one
matching I<rafsi>.

=item B<-R>

Attempt to match B<only> I<rafsi> and display only those I<valsi> that have at
least one matching I<rafsi>.

=item B<-w>

Only display matches that match on a full word, i.e., act as if there is a
C<\b> at either end of each operand.

=back

=head2 Handling Mutually Exclusive Options

Due to Getopt::Std's parsing of all options at once, in contrast to the "one at
a time" method of L<getopt(3)>, mutually exclusive options cannot be resolved
by checking which one came last; instead, an arbitrary set of precedences is
used:

=over

=item

The B<-C> and B<-c> options override the B<-G> and B<-g> options.

=item

B<-C>, B<-G>, and B<-R> override B<-c>, B<-g>, and B<-r>, respectively.

=item

If both B<-d> and B<-k> are given, they are treated as if just B<-e> had been
supplied.

=item

If B<-e> and either B<-d> or B<-k> are given, the B<-e> is ignored.

=item

If B<-d>, B<-e>, or B<-k> is given, B<-r> and B<-R> are ignored.

=item

The B<-q> option overrides the B<-a> option.

=back

=head1 PATTERN MATCHING

Treatment of operands is handled using a rather straightforward rule: if an
operand contains any special regex characters, it is treated as a Perl regular
expression; otherwise, it is treated as a literal string that the text being
compared against must be equal to or (if the B<-F> option is given or
definitions are being searched) must contain.  As an exception, parentheses are
not recognized as special characters because, due to the lack of capturing in
B<sisku>, they serve no use unless other special characters are present.  For
reference, the special regex characters recognized are:

    . * + ? ^ $ | [ ] { } \

If you wish to include one of these characters in a search pattern without
giving it any special meaning, you must precede it with a backslash (possibly
two backslashes, depending on your shell & use of quotes).  For more
information on L<perl(1)>'s regular expressions, run C<perldoc perlrequick>,
C<perldoc perlretut>, and/or C<perldoc perlre>.  If you supply a bad regex to
B<sisku>, whatever happens is your own fault.

When not using any of the B<-dek> options, 'h's in the operands are converted
to apostrophes before searching.

Note that any periods at the beginning of I<cmavo> in the I<cmavo> list are
ignored both when using regexes & when using literal strings.  Thus, the
operand C<.e> will match I<be> but not I<.e>, and forcing literal
interpretation with C<\.e> will cause the query to return no results.  Periods
elsewhere in I<cmavo> are left intact, but note that I<na.a> is the only entry
in the official I<cmavo> list to have a non-leading period.

=head1 RESTRICTIONS

B<sisku> does not search the notes or hints sections of entries.

Some I<cmavo> which appear in both the I<gismu> list and the I<cmavo> list have
different definitions in each, and currently the best way to handle them is to
concatenate them with "C< --- >".  This can generate odd-looking results,
especially for I<cmavo> whose two definitions differ only in a single word.

=head1 SEE ALSO

L<jvozba(1)>, Lojban::Valsi, Lojban::Vlasisku

I<The Complete Lojban Language> by John Woldemar Cowan

L<http://www.lojban.org/>

=head1 AUTHOR

John T. "kamymecraijun." Wodder II <minimiscience@gmail.com>

=head1 LICENSE

Feel free to do whatever the I<bais.> you want with this.

=cut
